### ISSUE TITLE
The inferred return type of a function with a generic that spreads an `import * as` doesn't include some exports when they're exported seperately to the declaration

### ISSUE THREAD

**Issue by mitchellhamilton** [microsoft/TypeScript#44179]
# Bug Report

### 🔎 Search Terms

export, import as, generic, spread

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about this and none seem to relate to this behavior 

### ⏯ Playground Link

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?#code/PTAEAEDMEsBsFMB2BDAtvAXKAzgFwK6SQB0u2AUOSDgO7IAO90iA5qLgBbyiT6IDGuaAHtEoACbx+sZACdkQ0aBrRO7Ltm6z42fLDKhm67tjTd4s2cNlUw05Nmw5h6TszYBvAL63Q-UXjOrhzuoAC8oB4AHhi4svjwXgDclNSoyADWoaqgyKD0stCoqtAAbtziwjqIAOS4ftoK3MhiFlay6sL4LBy+-oiB2C7wbqzhoDU1KeS8AopiQ8HuABQAlJE+5PBR9Nb1HkEjIWPJlNu7svWzgiKtO3vw4gAi0NqCsACeaxupYFBwSDMWGYkiipAoRQu9QAVLknHhCJAeFZUKAAETEYAIoho6bUGjWDIUbEkRZHdwpEnEc4PZ6vKS4T7Ta7zUAsEYAZQIRAAPAAVAB83w85FAfgC9VGbAiB2IcpJoFOYvxhIoYqlxDJUpSYqpNMujxeb0ZHx1oG0BFkYm15E21AAClZ6BZcB8JmSAGJ8fhPKSwGoSKpORDCerbaCBJSu50TA6e72+6QAUXuBrpxs+aywpWE0HESUVNWI5HZuC5iLWmuGNpVsiJJc53Mglf1uEN9PepqAA)

### 💻 Code

```ts repro
// @filename: stuff.ts

// swapping the function declaration with these results in the same error
// class something {}
// const something = {x:true};

// making it a primitive doesn't create an error though
// const something = '';

function something() {}

export { something };

export function exportedDirectly() {}

// @filename: index.ts
import * as stuff from "./stuff";

// works
stuff.something;
stuff.exportedDirectly;

function getStuff<T>() {
  const thing = { ...stuff };
  // works
  thing.something;
  stuff.exportedDirectly;
  return thing;
}

// Property 'something' does not exist on type '{ exportedDirectly(): void; }'.
getStuff().something;
// works
getStuff().exportedDirectly;
```

### 🙁 Actual behavior

Declarations that are exported separately to the declaration aren't included in the inferred return type with the exception of primitives (though maybe there are others that work as well?). This is wrong since the property exists at runtime and exists if the generic isn't there.

### 🙂 Expected behavior

All the exports are included since that's what happens at runtime and what happens when there isn't a generic on the function.

**Comment by RyanCavanaugh** [comment]
Wild. I'm not sure if ES6 module object exports are enumerated by spread or not, but at minimum the function body and its return type should behave the same.

**Comment by Zzzen** [comment]
> https://stackoverflow.com/a/38679368
> The important part of the spec in this case is that when you do
> 
>     import * as foo from 'foo';
> 
> The `foo` variable's value is created at [section 15.2.1.16.4 step 12.b](http://www.ecma-international.org/ecma-262/7.0/#sec-moduledeclarationinstantiation) which creates a [`Module Namespace Exotic Object`](http://www.ecma-international.org/ecma-262/7.0/#sec-module-namespace-exotic-objects), where [properties are the named exports](http://www.ecma-international.org/ecma-262/7.0/#sec-module-namespace-exotic-objects-ownpropertykeys) and [all properties are enumerable](http://www.ecma-international.org/ecma-262/7.0/#sec-module-namespace-exotic-objects-getownproperty-p) so you are totally safe in using `Object.keys(foo)` to get the names of all of the named exports. The object is not iterable, so you will not be able to use an iterable spread, though you could use the proposed object spread syntax to copy the properties if you wanted. `Object.getOwnPropertyNames` should work fine too.

TLDR: They are enumerable.
